### PR TITLE
feat: tracing improvements

### PR DESCRIPTION
**Pull request by deansheather**
- Move the tracing middleware higher up the stack so it can see more, but make sure that it only applies the backend API routes
- Add `tracing.RunWithoutSpan` which runs the given function with a noop span attached to the context. This is used in a few places to avoid capturing logs as events in the span when the span already had all of the information contained in the log.

### CODE REVIEW

File: coderd/coderd.go [hunk 1]
```diff
@@ -202,8 +202,9 @@ func New(options *Options) *API {
 	})
 
 	r.Use(
-		httpmw.AttachRequestID,
 		httpmw.Recover(api.Logger),
+		tracing.Middleware(api.TracerProvider),
```
Comment (kylecarbs) [inline coderd/coderd.go:206]: This will trace _every_ static HTTP file... is that ok?

Comment (deansheather) [reply coderd/coderd.go:206]: There's a filter within the middleware itself (and a test) to ensure that this is skipped on non-API routes.

File: coderd/httpmw/patternmatcher/routepatterns_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+func Test_RoutePatterns(t *testing.T) {
+	t.Parallel()
+
+	cases := []struct {
+		name        string
+		patterns    []string
+		errContains string
+		output      string
+	}{
```
Comment (mtojek) [inline coderd/httpmw/patternmatcher/routepatterns_test.go:19]: You may want to include more cases:
* `/`
* `//`
* `/api/`
* /api//v1`

I know that patterns are hard coded, but if we decide to generate them dynamically, some bugs in the future might be harder to investigate.

File: coderd/httpmw/logger.go [hunk 3]
```diff
@@ -11,9 +13,13 @@ import (
...
 func Logger(log slog.Logger) func(next http.Handler) http.Handler {
 	return func(next http.Handler) http.Handler {
-		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
 			start := time.Now()
-			sw := &tracing.StatusWriter{ResponseWriter: w}
+
+			sw, ok := rw.(*tracing.StatusWriter)
+			if !ok {
+				panic(fmt.Sprintf("ResponseWriter not a *tracing.StatusWriter; got %T", rw))
```
[the remark this answers is not in the document]
Comment (deansheather) [reply coderd/httpmw/logger.go:21]: The response writer will always be a `*tracing.StatusWriter`, so I think it's fine for us to panic here as this is a developer error. We panic in a other places due to developer errors as well, like the middleware context value stuff also in the httpmw package.

We also already do this exact panic in the tracing middleware too

[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/httpmw/logger.go:21]: I admit that I'm not a fan of it, but I understand why it's safe here 👍

**Review by coadler — APPROVED** [review]

**Pull request merged**